configfile: "config.yml"

SAMPLE = config["sample"]
REF = config["reference"]["fasta"]

rule all:
    input:
        "results/vcf/{sample}.filtered.vcf.gz".format(sample=SAMPLE),
        "results/multiqc"

# -----------------------------
# Download FASTQ
# -----------------------------
rule download_fastq:
    output:
        r1 = "samples/{sample}_1.fastq",
        r2 = "samples/{sample}_2.fastq"
    log:
        "logs/download_fastq/{sample}.log"
    shell:
        """
        mkdir -p logs/download_fastq
        wget -O {output.r1}.gz https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR258/004/{wildcards.sample}/{wildcards.sample}_1.fastq.gz
        wget -O {output.r2}.gz https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR258/004/{wildcards.sample}/{wildcards.sample}_2.fastq.gz
        gunzip -f {output.r1}.gz {output.r2}.gz &> {log}
        """

# -----------------------------
# Download reference genome
# -----------------------------
rule download_reference:
    output:
        REF
    log:
        "logs/reference_download.log"
    shell:
        """
        mkdir -p ref_genome logs
        wget -O {output}.gz {config[reference][url]}
        gunzip -f {output}.gz &> {log}
        """

rule bwa_index:
    input:
        REF
    output:
        amb = REF + ".amb",
        ann = REF + ".ann",
        bwt = REF + ".bwt",
        pac = REF + ".pac",
        sa  = REF + ".sa"
    threads: config["threads"]
    log:
        "logs/bwa_index.log"
    shell:
        "bwa index {input} &> {log}"

# -----------------------------
# FastQC
# -----------------------------
rule fastqc:
    input:
        r1 = config["fastq"]["r1"],
        r2 = config["fastq"]["r2"]
    output:
        directory("results/fastqc")
    threads: config["threads"]
    log:
        "logs/fastqc.log"
    shell:
        """
        mkdir -p results/fastqc logs
        fastqc -t {threads} -o {output} {input.r1} {input.r2}  &> {log}
        """

# -----------------------------
# Trimming
# -----------------------------
rule trim_reads:
    input:
        r1 = config["fastq"]["r1"],
        r2 = config["fastq"]["r2"]
    output:
        r1 = "results/trimmed/{sample}_1.trim.fastq",
        r2 = "results/trimmed/{sample}_2.trim.fastq"
    threads: config["threads"]
    log:
        "logs/trimmomatic/{sample}.log"
    shell:
        """
        mkdir -p logs/trimmomatic
        trimmomatic PE -threads {threads} \
        {input.r1} {input.r2} \
        {output.r1} /dev/null \
        {output.r2} /dev/null \
        SLIDINGWINDOW:4:20  &> {log}
        """

# -----------------------------
# Alignment
# -----------------------------
rule bwa_mem:
    input:
        ref = REF,
        idx = rules.bwa_index.output,
        r1 = "results/trimmed/{sample}_1.trim.fastq",
        r2 = "results/trimmed/{sample}_2.trim.fastq"
    output:
        "results/bam/{sample}.sorted.bam"
    threads: config["threads"]
    log:
        "logs/bwa_mem/{sample}.log"
    shell:
        """
        mkdir -p results/bam logs/bwa_mem
        bwa mem -t {threads} {input[0]} {input.r1} {input.r2} 2>> {log} |
        samtools sort -o {output}
        """

rule bam_index:
    input:
        "results/bam/{sample}.sorted.bam"
    output:
        "results/bam/{sample}.sorted.bam.bai"
    log:
        "logs/bam_index/{sample}.log"
    shell:
        """
        mkdir -p logs/bam_index
        samtools index {input} >> {log} 2>&1
        """

# -----------------------------
# Alignment QC (flagstat)
# -----------------------------
rule flagstat:
    input:
        "results/bam/{sample}.sorted.bam"
    output:
        "results/bam/{sample}.flagstat.txt"
    log:
        "logs/flagstat/{sample}.log"
    shell:
        """
        mkdir -p logs/flagstat
        samtools flagstat {input} > {output} 2> {log}
        """

# -----------------------------
# Variant Calling
# -----------------------------
rule mpileup:
    input:
        bam = "results/bam/{sample}.sorted.bam",
        ref = REF
    output:
        "results/bam/{sample}.bcf"
    log:
        "logs/bcftools_mpileup/{sample}.log"
    shell:
        "bcftools mpileup -O b -f {input.ref} {input.bam} -o {output} >> {log} 2>&1"

rule call_variants:
    input:
        "results/bam/{sample}.bcf"
    output:
        "results/vcf/{sample}.vcf"
    log:
        "logs/bcftools_call/{sample}.log"
    shell:
        "bcftools call --ploidy 1 -m -v -o {output} {input} >> {log} 2>&1"

# -----------------------------
# Variant Filtering
# -----------------------------
rule filter_variants:
    input:
        "results/vcf/{sample}.vcf"
    output:
        "results/vcf/{sample}.filtered.vcf.gz"
    log:
        "logs/variant_filter/{sample}.log"
    shell:
        """
        bcftools filter -i 'QUAL>30 && DP>10' {input} |
        bgzip -c > {output}
        tabix -p vcf {output} >> {log} 2>&1
        """

# -----------------------------
# MultiQC
# -----------------------------
rule multiqc:
    input:
        fastqc = directory("results/fastqc"),
        flagstat = expand("results/bam/{sample}.flagstat.txt", sample=[SAMPLE])
    output:
        directory("results/multiqc")
    log:
        "logs/multiqc.log"
    shell:
        "multiqc {input} -o {output} >> {log} 2>&1"
